src/heroic.rs

```diff
@@ -0,0 +1,252 @@
+    config::RootsConfig,
+    manifest::{Manifest, Store},
+    prelude::{normalize_title, StrictPath},
+};
+
+// Deserialization of Heroic gog_store/installed.json
+#[derive(serde::Deserialize)]
+struct HeroicInstalledGame {
+    #[serde(rename = "appName")]
+    game_id: String,
+    platform: String,
+}
+#[derive(serde::Deserialize)]
+struct HeroicInstalled {
+    installed: Vec<HeroicInstalledGame>,
+}
+
+// Deserialization of Heroic gog_store/library.json
+#[derive(serde::Deserialize)]
+struct GogLibraryGame {
+    app_name: String,
+    title: String,
+}
+#[derive(serde::Deserialize)]
+struct GogLibrary {
+    games: Vec<GogLibraryGame>,
+}
+
+// Deserialization of Legendary legendary/installed.json
+#[derive(serde::Deserialize)]
+struct LegendaryInstalledGame {
+    #[serde(rename = "app_name")]
+    game_id: String,
+    title: String,
+    platform: String,
+}
+#[derive(serde::Deserialize)]
+struct LegendaryInstalled(HashMap<String, LegendaryInstalledGame>);
+
+// TODO.2022-10-10 heroic: windows location for legendary
+// TODO.2022-10-15 heroic: make relative to heroic root in question!
+const LEGENDARY_PATHS: &[&str] = &[
+    "~/.config/legendary",
+    // TODO.2022-10-20 heroic: flatpak install is not supported yet
+    // "~/.var/app/com.heroicgameslauncher.hgl/config/legendary",
+];
+
+#[derive(Clone, Default, Debug)]
+pub struct HeroicGames {
+    games: HashMap<String, StrictPath>,
+    normalized_to_official: HashMap<String, String>,
+}
+
+impl HeroicGames {
+    pub fn get(&self, game: &String) -> Option<&StrictPath> {
+        self.games.get(game)
+    }
+
+    pub fn scan(roots: &[RootsConfig], manifest: &Manifest) -> Self {
+        let mut instance = HeroicGames {
+            normalized_to_official: manifest
+                .0
+                .keys()
+                .map(|title| (normalize_title(title), title.clone()))
+                .collect(),
+            ..Default::default()
+        };
+
+        roots.iter().for_each(|root: &RootsConfig| {
+            if root.store == Store::HeroicConfig {
+                instance.detect_flatpak_roots(root);
+                instance.detect_legendary_roots(root);
+                instance.detect_gog_roots(root);
+                log::trace!("detect_heroic_roots found: {:#?}", instance.games);
+            }
+        });
+
+        instance
+    }
+
+    fn detect_flatpak_roots(&mut self, _root: &RootsConfig) {
+        // TODO.2022-10-15 heroic: handle games storing stuff in ~/.var/app/com.heroicgameslauncher.hglw
+        // TODO.2022-10-15 heroic: for some games, .var/app/com.heroicgameslauncher.hglw is a root just like $HOME
+    }
+
+    fn detect_legendary_roots(&mut self, root: &RootsConfig) {
+        log::trace!("detect_legendary_roots searching for legendary config...");
+
+        for &legendary_path_candidate in LEGENDARY_PATHS {
+            let legendary_path = StrictPath::new(legendary_path_candidate.to_string());
+            if legendary_path.is_dir() {
```

Comment:
One trick for flattening the code is to do an early return/continue on the error case. That way, you can turn this:

```
if legendary_path.is_dir() {
    // indent grows
} else {
    // log
}
```

...into this:

```
if !legendary_path.is_dir() {
    // log
    continue;
}

// indent doesn't grow
```

When you have several error checks chained together like this, it makes the code easier to read.

Reply:
It [does indeed](https://enterprisecraftsmanship.com/posts/early-exit-is-tail-call-optimization/)